[dynamo] Register einops functions lazily

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110575
* #110990

Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

peterbell10 committed:
[dynamo] Register einops functions lazily

Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

[ghstack-poisoned]

peterbell10 committed:
Update on "[dynamo] Register einops functions lazily"


Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 committed:
Update on "[dynamo] Register einops functions lazily"


Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 committed:
Update on "[dynamo] Register einops functions lazily"


Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@peterbell10 is there anything I can do to help move this along? @Chillee was mentioning that a lot of people were running into this, and I'd like to get this in for the 2.1.1 patchfix release.

peterbell10 committed:
Update on "[dynamo] Register einops functions lazily"


Fixes #110549

We currently have a circular import between dynamo and einops as described in the issue.
This works around the issue by adding a mechanism to register initialization callbacks
that are called the first time an object is seen from that particular module.

This means that dynamo will only import `einops` after it's already fully initialized
and being called in a function being traced.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

atalman commented:
@peterbell10 could you please post cherry-pick for this PR in 2.1.1 tracker ?
